src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -684,7 +703,7 @@ private void ProduceStart(bool appCompleted)
 
         protected Task TryProduceInvalidRequestResponse()
         {
-            if (_requestRejected)
+            if (_requestRejected || _applicationException != null)
```

Comment:
If the response hasn't already started by the time we detect a mismatch in the content length, we can send a 500 response. `_applicationException` is set when there's a mismatch, so we need this check to run the error response logic.